### PR TITLE
ci: build GreptimeDB binary for later use

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Reuse the built GreptimeDB binary in CI. Let's see if it can save some time and money.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

**Review by zyy17 — APPROVED** [review]
LGTM

**Comment by zyy17** [comment]
@MichaelScofield Still have the error:

```
Caused by:
  process didn't exit successfully: `/greptimedb/target/debug/build/servers-18d50f6b6f3fd04d/build-script-build` (exit status: 101)
  --- stdout
  cargo:rustc-env=RUSTC_VERSION=rustc 1.76.0-nightly (3f28fe133 2023-12-18)

  --- stderr
  thread 'main' panicked at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/build-data-0.1.5/src/lib.rs:630:64:
  called `Result::unwrap()` on an `Err` value: "command 'git rev-parse --abbrev-ref=loose HEAD' failed: exit=128 stdout='' stderr='fatal: detected dubious ownership in repository at \\'/greptimedb\\'\\nTo add an exception for this directory, call:\\n\\n\\tgit config --global --add safe.directory /greptimedb\\n'"
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
make: *** [Makefile:66: build] Error 101
make: *** [Makefile:70: build-by-dev-builder] Error 2
```

**Comment by WenyXu** [comment]
We can use a small size for sqlness tests. Now, 4-CPU machines [are free for open-source projects](https://github.blog/2024-01-17-github-hosted-runners-double-the-power-for-open-source/). 
BTW, it could reduce compiling time if we introduce the cache. The `Build binaries` take 8 minutes(6min compiling, 2 min uploading) to finish the step, which is 2x the usage of the sqlness test.

**Comment by waynexia** [comment]
Also the kafka and normal sqlness cases can be run in the same job

File: tests/runner/src/env.rs [hunk 1]
```diff
@@ -79,18 +84,20 @@ impl EnvController for Env {
 
 impl Env {
     pub fn new(data_home: PathBuf, server_addr: Option<String>, wal: WalConfig) -> Self {
+        let bins_dir = std::env::var("GREPTIME_BINS_DIR").map(PathBuf::from).ok();
```
Comment (waynexia) [inline tests/runner/src/env.rs:87]: Why get this from env but not cli args?

**Comment by WenyXu** [comment]
> Also the kafka and normal sqlness cases can be run in the same job

BTW, I switch the sqlness runner to 4 core, Now it's free. maybe doesn't really matter.🥰

**Comment by MichaelScofield** [comment]
@zyy17 @WenyXu @waynexia PTAL

**Review by WenyXu — APPROVED** [review]

**Pull request merged**